### PR TITLE
[BEAM-4766] Migrate to use vendored protobuf for bytestring

### PR DESCRIPTION
**Pull request by lukecwik**
All existing usages were private to the class and this will reduce the size of the jar from ~6mbs to ~4.5mbs

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @apilloud @suztomo

**Comment by iemejia** [comment]
:clap:

**Comment by iemejia** [comment]
Maybe worth to add an extra checkstyle validation like the one we have for guava for protobuf to prevent accidental usage (if leaked from some dep)
https://github.com/apache/beam/blob/67ad5ce51d551f395d3a7a29b94c1eb0dc652b3a/sdks/java/build-tools/src/main/resources/beam/checkstyle.xml#L103

**Comment by suztomo** [comment]
I’m against vendoring in general. But I guess you have some problem in your mind. Would you be willing to share the problem?

**Comment by suztomo** [comment]
I forgot t check BEAM-4766

File: sdks/java/core/build.gradle [hunk 1]
```diff
@@ -69,7 +67,6 @@ dependencies {
   shadow library.java.vendored_bytebuddy_1_10_8
   shadow library.java.vendored_guava_26_0_jre
   compile library.java.antlr_runtime
-  compile library.java.protobuf_java
```
Comment (suztomo) [inline sdks/java/core/build.gradle:72]: Where does this module get the vendored protobuf dependency?

**Comment by lukecwik** [comment]
> [BEAM-4766](https://issues.apache.org/jira/browse/BEAM-4766) explains the concern in which protobuf's ByteString may be overridden by some other dependency.
> 
> I see the fields are private. There shouldn't be a problem.
> 
> Removing a dependency may reveal hidden diamond dependency conflicts. Did you run the linkage checker?

I didn't run the linkage checker because the dependency was never part of the dependency tree that is produced by the POMs but I can double check.

File: sdks/java/core/build.gradle (hunk 1 above)
Comment (lukecwik) [reply sdks/java/core/build.gradle:72]: It is a transitive dependency of `:model:pipeline`. Will update to depend on explicitly.

**Comment by lukecwik** [comment]
Linkage checker results:
```
beam-linkage-check.sh "beam-sdks-java-core"
remote: Enumerating objects: 8, done.
remote: Counting objects: 100% (8/8), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 8 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (8/8), 3.48 KiB | 890.00 KiB/s, done.
From https://github.com/apache/beam
   fcb40cfa48..9c16b898f0  master     -> origin/master
Previous HEAD position was e0def937b1 fixup! Address PR comments
HEAD is now at 9c16b898f0 Merge pull request #11890 from kennknowles/jira-priorities
Tue 02 Jun 2020 10:21:17 AM PDT: Installing artifacts of master(9c16b898) to Maven local repository.
Tue 02 Jun 2020 10:25:01 AM PDT: Running linkage check (baseline) for beam-sdks-java-core in master
Tue 02 Jun 2020 10:25:01 AM PDT: to create a baseline (existing errors before change) build/linkagecheck/baseline-beam-sdks-java-core.xml
Configuration on demand is an incubating feature.

> Task :runners:apex:buildDependencyTree
See the report at: file:///usr/local/google/home/lcwik/git/beam/runners/apex/build/classes/java/main/org/apache/beam/runners/apex/dependency-tree

> Task :checkJavaLinkage
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Wrote the linkage errors as exclusion file: build/linkagecheck/baseline-beam-sdks-java-core.xml
NOTE: This task published artifacts into your local Maven repository. You may want to remove them manually.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.2.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 20s
968 actionable tasks: 172 executed, 796 up-to-date
Tue 02 Jun 2020 10:25:22 AM PDT: Installing artifacts of HEAD(e0def937) to Maven local repository.
Tue 02 Jun 2020 10:25:41 AM PDT: Running linkage check (validate) for beam-sdks-java-core in HEAD
Tue 02 Jun 2020 10:25:41 AM PDT: using baseline build/linkagecheck/baseline-beam-sdks-java-core.xml
Configuration on demand is an incubating feature.

> Task :runners:apex:buildDependencyTree
See the report at: file:///usr/local/google/home/lcwik/git/beam/runners/apex/build/classes/java/main/org/apache/beam/runners/apex/dependency-tree

> Task :checkJavaLinkage
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
NOTE: This task published artifacts into your local Maven repository. You may want to remove them manually.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.2.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 18s
968 actionable tasks: 172 executed, 796 up-to-date
Tue 02 Jun 2020 10:26:00 AM PDT: Done: 0
No new linkage errors
```

**Comment by iemejia** [comment]
Oups I have to lower my previous enthusiasm because I had not thought that putting explicitly the vendored grpc version in sdks/java/core brings some extra transitive 20M for `sdks/java/core` users for not much :face_with_head_bandage: Wondering if this was the right fix for this problem.

**Comment by iemejia** [comment]
I expected that vendored-grpc could be coming via some transitive dependency but it wasn't
https://repo1.maven.org/maven2/org/apache/beam/beam-sdks-java-core/2.21.0/beam-sdks-java-core-2.21.0.pom

Maybe we should revert this, the only positive outcome I can see of this now is if at some moment `sdks/java/core` wants to act explictly on gRPC services which I doubt save for the `External` transform but this transform can easily be an extension to avoid adding the 20M in core.

**Comment by iemejia** [comment]
Any comments on this @lukecwik?

**Comment by lukecwik** [comment]
beam-sdks-java-core depends on beam-model-pipeline which depends on beam-vendor-grpc-1_26_0 so what you initially thought was correct, was there an issue in how you dumped the transitive dependency tree?

**Comment by iemejia** [comment]
Oh I missed that one. I suppose this is ok. Thanks for the info. Our deps have become huge I really wish we could find a way to make them smaller but it seems we only can get some minor fixes back and ther but the bigger ones are hard to match.

**Pull request merged**